### PR TITLE
ddl:support add index rollback of partitioned table

### PR DESCRIPTION
**Pull request by ciscoxll**
### What problem does this PR solve? 

* This PR solves the problem of  partitioned table add index does not support `rollback`.
### What is changed and how it works?
* Handle the rolling back job and finish the PR (support drop index for the partitioned table) #7306 first.
* `Drop index` mainly deletes `indexInfo`.
* If add index job rollbacks in write reorganization state, its need to delete all keys which has been added.
* Delete all keys is done in the `delete range`.
*  Fix issue #7223

### Check List 

Tests 

 - Unit test

PTAL @tiancaiamao @winkyao @zimulala .

### CODE REVIEW

File: ddl/delete_range.go [hunk 1]
```diff
@@ -268,12 +268,23 @@ func insertJobIntoDeleteRangeTable(ctx sessionctx.Context, job *model.Job) error
 	case model.ActionAddIndex:
 		tableID := job.TableID
 		var indexID int64
-		if err := job.DecodeArgs(&indexID); err != nil {
+		var partitionIDs []int64
+		if err := job.DecodeArgs(&indexID, &partitionIDs); err != nil {
```
Comment (jackysp) [inline ddl/delete_range.go]: Will it break the backward compatibility?

Comment (ciscoxll) [reply ddl/delete_range.go]: @jackysp Just adding parameters does not break compatibility.

File: ddl/db_test.go [hunk 2]
```diff
@@ -3203,6 +3203,130 @@ LOOP:
...
+		// It only tests cancel one DDL job.
+		if errs[0] != nil {
+			checkErr = errors.Trace(errs[0])
+			return
+		}
+		err = hookCtx.Txn().Commit(context.Background())
+		if err != nil {
+			checkErr = errors.Trace(err)
+		}
+	}
```
Comment (zimulala) [inline ddl/db_test.go]: Is this function the same as the `hook.OnJobUpdatedExported` in `TestCancelAddIndex`?

**Comment by tiancaiamao** [comment]
LGTM

File: ddl/db_test.go (hunk 2 above)
Comment (ciscoxll) [reply ddl/db_test.go]: @zimulala Yes, they are the same.

File: ddl/delete_range.go [hunk 3]
```diff
@@ -268,12 +268,23 @@ func insertJobIntoDeleteRangeTable(ctx sessionctx.Context, job *model.Job) error
...
 		}
-		startKey := tablecodec.EncodeTableIndexPrefix(tableID, indexID)
-		endKey := tablecodec.EncodeTableIndexPrefix(tableID, indexID+1)
-		return doInsert(s, job.ID, indexID, startKey, endKey, now)
+		if len(partitionIDs) > 0 {
+			for _, pid := range partitionIDs {
+				startKey := tablecodec.EncodeTableIndexPrefix(pid, indexID)
+				endKey := tablecodec.EncodeTableIndexPrefix(pid, indexID+1)
+				if err := doInsert(s, job.ID, indexID, startKey, endKey, now); err != nil {
+					return errors.Trace(err)
```
Comment (XuHuaiyu) [inline ddl/delete_range.go]: Using `break` here can help us to save `else`.

File: ddl/db_test.go [hunk 4]
```diff
@@ -3203,6 +3203,130 @@ LOOP:
 	s.tk.MustExec("drop table partition_drop_idx;")
 }
 
+func (s *testDBSuite) TestPartitionCancelAddIndex(c *C) {
+	s.tk = testkit.NewTestKit(c, s.store)
+	s.mustExec(c, "use test_db")
+	s.mustExec(c, "set @@session.tidb_enable_table_partition=1;")
```
Comment (XuHuaiyu) [inline ddl/db_test.go]: should we set @@session.tidb_enable_table_partition=0; at the end of this function.

Comment (tiancaiamao) [reply ddl/db_test.go]: It's should be OK to `set @@session.tidb_enable_table_partition=1; ` for any test code. @XuHuaiyu 
We'll enable table partition by default later.

File: ddl/delete_range.go (hunk 3 above)
Comment (ciscoxll) [reply ddl/delete_range.go]: @XuHuaiyu It feels like it doesn't have to be modified, these are two different kinds of logic.

File: ddl/db_test.go (hunk 2 above)
Comment (ciscoxll) [reply ddl/db_test.go]: @zimulala Done.

**Comment by ciscoxll** [comment]
@zimulala  PTAL.

File: ddl/db_test.go [hunk 5]
```diff
@@ -3203,6 +3203,137 @@ LOOP:
...
+	idx := tables.NewIndex(pid, t.Meta(), c3IdxInfo)
+	checkDelRangeDone(c, ctx, idx)
+
+	s.mustExec(c, "drop table t1_cancel_add_index")
+	ddl.ReorgWaitTimeout = oldReorgWaitTimeout
+	callback := &ddl.TestDDLCallback{}
+	s.dom.DDL().(ddl.DDLForTest).SetHook(callback)
+}
+
+func backgroundExecOnJobUpdatedExported(c *C, s *testDBSuite, hook *ddl.TestDDLCallback, checkErr error) (func(*model.Job), *model.IndexInfo) {
```
Comment (zimulala) [inline ddl/db_test.go]: Cloud we use it in "TestCancelAddIndex"? And hold the comment in original “TestCancelAddIndex”.

Comment (ciscoxll) [reply ddl/db_test.go]: @zimulala Done.

**Pull request merged**